man/repl.man
.TH REPL %manext1% 2022-12-22 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
repl \- reply to an nmh message
.SH SYNOPSIS
.HP 5
.na
.B repl
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RI [ msg ]
.RB [ \-annotate " | " \-noannotate ]
.RB [ \-group " | " \-nogroup ]
.RB [ \-cc
all/to/cc/me]
.RB [ \-nocc
all/to/cc/me]
.RB [ \-query " | " \-noquery ]
.RB [ \-form
.IR formfile ]
.RB [ \-format " | " \-noformat ]
.RB [ \-filter
.IR filterfile ]
.RB [ \-inplace " | " \-noinplace ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-fcc
.IR +folder ]
.RB [ \-width
.IR columns ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-editor
.IR editor ]
.RB [ \-noedit ]
.RB [ \-convertargs
.IR "type argstring" ]
.RB [ \-whatnowproc
.IR program ]
.RB [ \-nowhatnowproc ]
.RB [ \-atfile ]
.RB [ \-noatfile ]
.RB [ \-fmtproc
.IR program ]
.RB [ \-nofmtproc ]
.RB [ \-build ]
.RB [ \-file
.IR msgfile ]
.ad
.SH DESCRIPTION
.B repl
may be used to reply to a message.
.PP
In its simplest form (with no arguments),
.B repl
will set up a message-form skeleton in reply to the current message
in the current folder, and invoke the
.B whatnow
shell.
.PP
.B repl
uses a reply template to construct the draft of the reply.
A reply template is simply an
.B mhl
format file (see
.IR mh\-format (5)
for details).
.PP
If the switch
.B \-nogroup
is given (it is on by default), then
.B repl
will use the standard forms file \*(lqreplcomps\*(rq.  This will construct
a draft message that is intended to be sent only to the author of the
message to which you are replying.  If a file named \*(lqreplcomps\*(rq
exists in the user's
.B nmh
directory, it will be used instead of this default forms file.
.PP
The default reply template \*(lqreplcomps\*(rq will direct
.B repl
to construct the reply message draft as follows:
.PP
.RS 5
.nf
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
cc: <To> and <cc> and <personal address>
Fcc: {fcc switch} or +outbox
Subject: Re: <Subject>
In\-Reply\-To: <Message\-Id>
References: <Message\-Id>
Comments: In\-Reply\-To <From> or <apparently from> or <Sender>
.ti +3
message dated <date>
--------
.fi
.RE
.PP
where field names enclosed in angle brackets (<\ >) indicate the
contents of the named field from the message to which the reply is
being made.
.PP
By default, the \*(lqcc:\*(rq field is empty.  You may selectively add
addresses to this default with the
.B \-cc
.I type
switch.  This switch takes an argument
.RI ( all / to / cc / me )
which specifies who gets added to the default
\*(lqcc:\*(rq list of the reply.  You may give this switch multiple times (with
different arguments) if you wish to add multiple types of address.
.PP
If the switch
.B \-group
is given, then
.B repl
will use the standard forms file \*(lqreplgroupcomps\*(rq.
This will construct a draft message that is intended as a group or
followup reply.  If a file named \*(lqreplgroupcomps\*(rq exists
in the user's nmh directory, it will be used instead of this default
forms file, unless you specify another forms file on the command line
or in your profile.
.PP
The default group reply template \*(lqreplgroupcomps\*(rq will direct
.B repl
to construct the reply message draft as follows:
.PP
.RS 5
.nf
To: <Mail\-Followup\-To>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
--------
.fi
.RE
.PP
or if the field <Mail\-Followup\-To> is not available:
.PP
.RS 5
.nf
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
cc: <To> and <cc> and <personal address>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
--------
.fi
.RE
.PP
By default, the \*(lqcc:\*(rq contains all the addresses shown.  You may
selectively remove addresses from this default with the
.B \-nocc
.I type
switch.
This switch takes an argument (
.IR all / to / cc / me )
which specifies who gets removed from the default \*(lqcc:\*(rq list of
the reply.  You may give this switch multiple times (with different arguments)
if you wish to remove multiple types of address.
.PP
In any case, you may specify an alternative forms file with the switch
.B \-form
.IR formfile .
.PP
The
.B \-query
switch modifies the action of
.B \-nocc
.I type
switch by interactively asking you if each address that normally would be
placed in the \*(lqTo:\*(rq and \*(lqcc:\*(rq list should actually be sent a copy.
This is useful for special-purpose replies.  Note that the position of
the
.B \-cc
and
.B \-nocc
switches, like all other switches which take a positive and negative form,
is important.
.PP
Lines beginning with the fields \*(lqTo:\*(rq, \*(lqcc:\*(rq, and
\*(rqBcc:\*(rq will be standardized and have duplicate addresses removed.
In addition, the
.B \-width
.I columns
switch will guide
.BR repl 's
formatting of these fields.
.PP
If the draft already exists,
.B repl
will ask you as to the disposition of the draft.  A reply of
.B quit
will abort
.BR repl ,
leaving the
draft intact;
.B replace
will replace the existing draft with a blank skeleton; and
.B list
will display the draft.
.PP
See
.IR comp (1)
for a description of the
.B \-editor
and
.B \-noedit
switches.  Note that while in the editor, with
.B \-atfile
and if the current directory is writable, the message being replied
to is available through a link named \*(lq@\*(rq (assuming the default
.IR whatnowproc ).
In addition, the actual pathname of the message is stored in the
environment variable
.BR $editalt ,
and the pathname of the folder containing the message is stored in
the environment variable
.BR $mhfolder .
The creation of the \*(lq@\*(rq file is controlled via the
.B \-atfile
and
.B \-noatfile
options.
.PP
The
.B \-convertargs
switch directs
.B repl
to pass the arguments for
.I type
to
.BR mhbuild .
Both arguments are required;
.I type
must be non-empty while
.I argstring
can be empty, e.g., '' in a shell command line.  The
.B \-convertargs
switch can be used multiple times.
See the Convert Interface section of
.IR mhbuild (1)
for a description of the convert mechanism, and
.I %docdir%/contrib/replaliases
for examples of its use.
.PP
Although
.B repl
uses a forms file to direct it how to construct the beginning of the
draft, it uses a message filter file to direct it as to how the message
to which you are replying should be filtered (re-formatted) in the body
of the draft.  The filter file for
.B repl
should be a standard form file for
.BR mhl ,
as
.B repl
will invoke
.B mhl
to format the message to which you are replying.
.PP
The switches
.BR \-noformat ,
.BR \-format ,
and
.B \-filter
.I filterfile
specify which message filter file to use.
.PP
If the switch
.B \-noformat
is given (it is the default) and the
.B \-filter
switch is not used, then the message to which you are replying is not
included in the body of the draft.
.PP
If the switch
.B \-format
is given, then a default message filter file is used.
This default message filter should be adequate for most users.
This default filter
.RI \*(lq mhl.reply \*(rq
is:
.PP
.RS 5
.nf
%mhl_reply%
.fi
.RE
.PP
which outputs each line of the body of the message prefaced with the
\*(lq>\*(rq character and a space.
.PP
If a file named
.RI \*(lq mhl.reply \*(rq
exists in the user's
.B nmh
directory,
it will be used instead of this form.  You may specify an alternate
message filter file with the switch
.B \-filter
.IR filterfile .
.PP
Other reply filters are commonly used, such as:
.PP
.RS 5
.nf
:
body:nocomponent,compwidth=9,offset=9
.fi
.RE
.PP
which says to output a blank line, and then the body of the message
being replied to, indented by one tab stop.  Another popular format
is:
.PP
.RS 5
.nf
message-id:nocomponent,\|nonewline,\\
formatfield=\*(lqIn message %{text},\ \*(rq
from:nocomponent,\|formatfield=\*(lq%(decode(friendly{text})) writes:\*(rq
body:component=\*(lq>\*(rq,\|overflowtext=\*(lq>\*(rq,\|overflowoffset=0
.fi
.RE
.PP
This message filter file cites the Message-ID and author of the message
being replied to, and then outputs each line of the body prefaced with
the \*(lq>\*(rq character.
.PP
You can also use an external format program to format the message body.
The format program is specified by the
.I formatproc
profile entry, and is enabled by the \*(lqformat\*(rq flag.  A message
filter using an external format program would look like this:
.PP
.RS 5
.nf
body:component=\*(lq>\*(rq,\|nowrap,\|format
.fi
.RE
.PP
See the
.IR mhl (1)
documentation for more information.
The format program can be changed by the
.B \-fmtproc
.I program
and
.B \-nofmtproc
switches.
.PP
To use the MIME rules for encapsulation, specify the
.B \-mime
switch.
This directs
.B repl
to generate an
.B mhbuild
composition file.
Note that
.B nmh
will not invoke
.B mhbuild
automatically; you must specifically give the command
.PP
.RS 5
What now? mime
.RE
.PP
prior to sending the draft.
.PP
If the
.B \-annotate
switch is given, the message being replied to will be annotated with
the lines
.PP
.RS 5
Replied:\ date
Replied:\ addrs
.RE
.PP
where the address list contains one line for each addressee.
The annotation will be done only if the message is sent directly from
.BR repl .
If the message is not sent immediately from
.BR repl ,
.RB \*(lq "comp\ \-use" \*(rq
may be used to re-edit and send the constructed message, but the
annotations won't take place.
Normally annotations are done inplace in order to preserve any links
to the message.
You may use the
.B \-noinplace
switch to change this.
.PP
Although the default template specifies that a copy of the reply will be
put in the folder `outbox', if the
.B \-fcc
.I +folder
switch is given it will override the default value.
More than one folder, each preceded by
.B \-fcc
can be named.
.PP
In addition to the standard
.IR mh\-format (5)
escapes,
.B repl
also recognizes the following additional
.I component
escape:
.PP
.RS 5
.nf
.ta \w'Escape  'u +\w'Returns  'u
.I "Escape	Returns	Description"
fcc	string	Any folders specified with `\-fcc\ folder'
.fi
.RE
.PP
To avoid reiteration,
.B repl
strips any leading `Re: ' strings from the
.I subject
component.
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke the
.B nmh
draft folder facility.  This is an advanced (and highly useful) feature.
Consult
.IR mh-draft (5)
for more information.
.PP
Upon exiting from the editor,
.B repl
will invoke the
.B whatnow
program.  See
.IR whatnow (1)
for a discussion of available
options.  The invocation of this program can be inhibited by using the
.B \-nowhatnowproc
switch.  (In fact, it is the
.B whatnow
program which starts the initial edit.  Hence,
.B \-nowhatnowproc
will prevent any edit from occurring.)
.PP
The
.B \-build
switch is intended to be used by the Emacs mh-e interface
to
.BR nmh .
It implies
.BR \-nowhatnowproc .
It causes a file
.I <mh-dir>/reply
to be created, containing the draft message that would normally be presented
to the user for editing.
No mail is actually sent.
.PP
The
.B \-file
.I msgfile
switch specifies the message to be replied to as an exact filename rather
than as an
.B nmh
folder and message number.
The same caveats apply to this option as to the
.B \-build
switch.
.SH FILES
.B repl
looks for all format, filter and template files in multiple locations:
absolute pathnames are accessed directly, tilde expansion is done on
usernames, and files are searched for in the user's
.I Mail
directory as specified in their profile.  If not found there, the directory
.RI \*(lq %nmhetcdir% \*(rq
is checked.
.PP
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^%nmhetcdir%/replcomps~^The standard reply template
^or <mh-dir>/replcomps~^Rather than the standard template
^%nmhetcdir%/replgroupcomps~^The standard `reply -group' template
^or <mh-dir>/replgroupcomps~^Rather than the standard template
^%nmhetcdir%/mhl.reply~^The standard message filter
^or <mh-dir>/mhl.reply~^Rather than the standard filter
^$HOME/.mh_profile~^The user profile
^<mh-dir>/draft~^The draft file
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Alternate\-Mailboxes:~^To determine the user's mailboxes
^Current\-Folder:~^To find the default current folder
^Draft\-Folder:~^To specify the default draftfolder
^Editor:~^To override the default editor
^Msg\-Protect:~^To set mode when creating a new message (draft)
^fileproc:~^Program to refile the message
^mhlproc:~^Program to filter message being replied-to
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.fi
.SH "SEE ALSO"
.IR comp (1),
.IR forw (1),
.IR mh\-draft (5),
.IR mh\-format (5),
.IR mhbuild (1),
.IR send (1),
.IR whatnow (1)
.PP
.I %docdir%/contrib/replaliases
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder"
.RB ` msg "' defaults to cur"
.RB ` \-nogroup '
.RB ` "\-nocc\ all" "' with `\-nogroup', `\-cc\ all' with `\-group'"
.RB ` \-noannotate '
.RB ` \-nodraftfolder '
.RB ` \-noformat '
.RB ` \-inplace '
.RB ` \-nomime '
.RB ` \-noquery '
.RB ` \-noatfile '
.RB ` "\-width\ 72" '
.fi
.SH CONTEXT
If a folder is given, it will become the current folder.  The message
replied to will become the current message.
.SH BUGS
If any addresses occur in the reply template, addresses in the template
that do not contain hosts are defaulted incorrectly.  Instead of using
the localhost for the default,
.B repl
uses the sender's host.
Moral of the story: if you're going to include addresses in a reply
template, include the host portion of the address.
.PP
The
.B \-width
.I columns
switch is only used to do address-folding; other headers are not line-wrapped.
.PP
If
.I whatnowproc
is
.BR whatnow ,
then
.B repl
uses a built-in
.BR whatnow ,
it does not actually run the
.B whatnow
program.
Hence, if you define your own
.IR whatnowproc ,
don't call it
.B whatnow
since
.B repl
won't run it.


NEWS
#
# NEWS - Release notes for nmh 1.8
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly five years since the last release of nmh, and there have
been a number of significant changes since that last release.  Long-time
MH and nmh users should read carefully the NOTABLE CHANGES section, as
there are some significant changes to nmh behavior.  Otherwise, please
see the README and INSTALL files for help on getting started with nmh.

This release is dedicated to Norman Z. Shapiro, co-designer of the MH
Message Handling System.  MH is the predecessor of nmh.  Norm was an
active supporter of nmh development until he passed away in October of
2021.  We are most grateful to Norm for his stewardship of MH and nmh.
    https://en.wikipedia.org/wiki/Norman_Shapiro

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

- Support for Content-MD5 header fields, MIME content cache functionality,
  and the message/partial MIME type have been removed.
- Gmail OAuth2/XOAUTH support for desktop applications has been effectively
  dropped, so nmh no longer supports it.  nmh support for Gmail API access
  is experimental, please post to nmh-workers@nongnu.org if you'd like to
  help with test and development.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.  Because encoding can wrap
  long lines, the use of a paragraph formatter has been removed from
  mhn.defaults.

------------
NEW FEATURES
------------

- The default editor has been changed from 'vi' to 'prompter', to align with
  historical practice and reduce packaging dependencies on external programs.
- A new -checkbase64 switch has been added to mhfixmsg(1).
- inc(1)/msgchk(1) now support STARTTLS for the POP protocol.
- All TLS-supported protocols now will send the SNI (server name indicator)
  TLS extension.
- A new mh-format function %(ordinal) has been implemented to output the
  appropriate ordinal suffix for numbers.  Example: "%(num 22)%(ordinal)"
  will output "22nd".
- show and mhl now decode more addresses in header fields.
- Added warning from all programs that read the profile if the profile
  contains a post entry, which is ignored, but does not contain a
  postproc entry.  In other words, if you get this warning and want
  to suppress it, your options include:
  1) Remove your post profile entry.
  2) Make your post profile entry a comment by prepending it with the #:
     comment indicator.
  3) Add a postproc entry that points to the post that you use.  That can
     be viewed with "mhparam postproc".
- scan(1) -file argument can be a Maildir directory.
- Updated mhn.defaults to prefer mpv(1) over xv(1) and replace mpeg_play(1),
  and to use it for all video types, not just video/mpeg.  And prefer all
  other searched-for pdf viewers over acroread(1).
- Added mhshow-suffix-video.mp4 to mhn.defaults, for use by mhshow(1) and
  send(1).
- Removed support from mhn.defaults for application/x-ivs and text/richtext.
- Changed interpretation of argument to mhfixmsg(1) -decodeheaderfieldbodies
  switch to specify character set of the decoded field bodies.
- repl(1) -convertargs now allows editing of the composition draft between
  translation and any encoding of text content.
- install-mh(1) now enables the mh-draft(5) draft folder facility.

-----------------
OBSOLETE FEATURES
-----------------

- The generation and verification of a Content-MD5 field has been removed
  without deprecation.  The related -check and -nocheck options now error.
- The MIME content cache functionality has been mostly non-functional since
  the development on nmh, and consequently all of the content caching code
  and related switches (-cache/-rcache/-wcache) have been removed.
- Support for generating and reassembling message/partial messages has been
  removed; it seems that this has been broken since 1.5 and there is very
  little support across MUAs.
- Marked Gmail OAuth2/XOAUTH support as being unsupported.
- Support for the MHPDEBUG environment variable was removed.  It was
  deprecated in nmh 1.7.  The pick(1) -debug switch replaced it.
- The 'libdir' mhparam(1) component was removed.  It was deprecated in
  nmh 1.7, when it was replaced by a new 'libexecdir' component.

---------
BUG FIXES
---------

- Fixed bcc to work with sendmail/pipe, and better documented that dcc
  doesn't work with it [Bug 55700].
- An -attendee switch has been added to mhical(1), for use when more than one
  (or zero) attendees match a user's mailbox.
- Fixed inc(1) and %(me) function escape to not obey Local-Mailbox profile
  component.
- Fixed source charset in mhfixmsg textcharset verbose output.
- Fixed mhfixmsg charset determination of content added with -reformat.
- Fixed file descriptor leak in mhfixmsg when run on multiple input files.
- Fixed mhfixmsg(1) -decodeheaderfilebodies to support mixed encoded/undecoded.
- Fixed memory corruption in post(1) and inc(1) when using XOAUTH2,
  with 4 or more entries in the oauth-authservice file.
- Added alias expansion to From: address for use by sendfrom.
- Removed extra space added before header field bodies by dist(1) to $mhdraft.
- Fixed display of iCalendar object with multiple VEVENTS using a VTIMEZONE.
- Fixed allowable encodings with MIME message types to get closer to RFC 2046.
- Detect other files regardless of backup prefix [Bug #49476].
- Copy if hard link by refile(1) or send(1) fails with EACESS [Bug #56575].

----------------------------------------------------------------------------

#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time

uip/repl.c
/* repl.c -- reply to a message
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include "h/mh.h"
#include "whatnowproc.h"
#include "annosbr.h"
#include "sbr/m_name.h"
#include "sbr/getarguments.h"
#include "sbr/read_switch_multiword.h"
#include "sbr/concat.h"
#include "sbr/seq_setprev.h"
#include "sbr/seq_setcur.h"
#include "sbr/seq_save.h"
#include "sbr/showfile.h"
#include "sbr/smatch.h"
#include "sbr/refile.h"
#include "sbr/m_draft.h"
#include "sbr/m_convert.h"
#include "sbr/getfolder.h"
#include "sbr/folder_read.h"
#include "sbr/context_save.h"
#include "sbr/context_replace.h"
#include "sbr/context_find.h"
#include "sbr/ambigsw.h"
#include "sbr/path.h"
#include "sbr/print_version.h"
#include "sbr/print_help.h"
#include "sbr/error.h"
#include "h/mime.h"
#include "sbr/done.h"
#include "sbr/utils.h"
#include "sbr/m_maildir.h"
#include "replsbr.h"
#include "sbr/globals.h"

#define REPL_SWITCHES \
    X("group", 0, GROUPSW) \
    X("nogroup", 0, NGROUPSW) \
    X("annotate", 0, ANNOSW) \
    X("noannotate", 0, NANNOSW) \
    X("cc all|to|cc|me", 0, CCSW) \
    X("nocc all|to|cc|me", 0, NCCSW) \
    X("draftfolder +folder", 0, DFOLDSW) \
    X("draftmessage msg", 0, DMSGSW) \
    X("nodraftfolder", 0, NDFLDSW) \
    X("editor editor", 0, EDITRSW) \
    X("noedit", 0, NEDITSW) \
    X("convertargs type argstring", 0, CONVERTARGSW) \
    X("fcc folder", 0, FCCSW) \
    X("filter filterfile", 0, FILTSW) \
    X("form formfile", 0, FORMSW) \
    X("format", 5, FRMTSW) \
    X("noformat", 7, NFRMTSW) \
    X("inplace", 0, INPLSW) \
    X("noinplace", 0, NINPLSW) \
    X("mime", 0, MIMESW) \
    X("nomime", 0, NMIMESW) \
    X("query", 0, QURYSW) \
    X("noquery", 0, NQURYSW) \
    X("whatnowproc program", 0, WHATSW) \
    X("nowhatnowproc", 0, NWHATSW) \
    X("width columns", 0, WIDTHSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \
    X("file file", 4, FILESW) \
    X("build", 5, BILDSW) /* interface from mhe */ \
    X("atfile", 0, ATFILESW) \
    X("noatfile", 0, NOATFILESW) \
    X("fmtproc program", 0, FMTPROCSW) \
    X("nofmtproc", 0, NFMTPROCSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(REPL);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(REPL, switches);
#undef X

#define CC_SWITCHES \
    X("to", 0, CTOSW) \
    X("cc", 0, CCCSW) \
    X("me", 0, CMESW) \
    X("all", 0, CALSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(CC);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(CC, ccswitches);
#undef X

#define DISPO_SWITCHES \
    X("quit", 0, NOSW) \
    X("replace", 0, YESW) \
    X("list", 0, LISTDSW) \
    X("refile +folder", 0, REFILSW) \
    X("new", 0, NEWSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(DISPO);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(DISPO, aqrnl);
#undef X

static struct swit aqrl[] = {
    { "quit", 0, NOSW },
    { "replace", 0, YESW },
    { "list", 0, LISTDSW },
    { "refile +folder", 0, REFILSW },
    { NULL, 0, 0 }
};

static short outputlinelen = OUTPUTLINELEN;
static bool groupreply;			/* Is this a group reply?        */

static bool mime;			/* include original as MIME part */
static char *form   = NULL;		/* form (components) file        */
static char *filter = NULL;		/* message filter file           */
static char *fcc    = NULL;		/* folders to add to Fcc: header */


/*
 * prototypes
 */
static void docc (char *, int);
static void add_convert_header (const char *, char *, char *, char *);


int
main (int argc, char **argv)
{
    int	i, isdf = 0;
    bool anot = false;
    bool inplace = true;
    bool nedit = false;
    bool nwhat = false;
    bool atfile = false;
    int fmtproc = -1;
    char *cp, *cwd, *dp, *maildir, *file = NULL;
    char *folder = NULL, *msg = NULL, *dfolder = NULL;
    char *dmsg = NULL, *ed = NULL, drft[BUFSIZ], buf[BUFSIZ];
    char **argp, **arguments;
    svector_t convert_types = svector_create (10);
    svector_t convert_args = svector_create (10);
    size_t n;
    struct msgs *mp = NULL;
    struct stat st;
    FILE *in;
    bool buildsw = false;

    if (nmh_init(argv[0], true, true)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    die("-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s: [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case GROUPSW:
		    groupreply = true;
		    continue;
		case NGROUPSW:
		    groupreply = false;
		    continue;

		case ANNOSW: 
		    anot = true;
		    continue;
		case NANNOSW: 
		    anot = false;
		    continue;

		case CCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    docc (cp, 1);
		    continue;
		case NCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    docc (cp, 0);
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			die("missing argument to %s", argp[-2]);
		    nedit = false;
		    continue;
		case NEDITSW:
		    nedit = true;
		    continue;
		    
		case CONVERTARGSW: {
                    char *type;
                    size_t i;

		    if (!(type = *argp++)) {
			die("missing type argument to %s", argp[-2]);
                    }
		    if (!(cp = *argp++)) {
			die("missing argstring argument to %s",
			       argp[-3]);
                    }

                    for (i = 0; i < svector_size (convert_types); ++i) {
                        if (! strcmp (svector_at (convert_types, i), type)) {
                            /* Already saw this type, so just update
                               its args. */
                            svector_strs (convert_args)[i] = cp;
                            break;
                        }
                    }

                    if (i == svector_size (convert_types)) {
                        svector_push_back (convert_types, type);
                        svector_push_back (convert_args, cp);
                    }
		    continue;
                }

		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			die("missing argument to %s", argp[-2]);
		    nwhat = false;
		    continue;
		case BILDSW: 
		    buildsw = true;
		    /* FALLTHRU */
		case NWHATSW: 
		    nwhat = true;
		    continue;

		case FCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dp = NULL;
		    if (*cp == '@')
			cp = dp = path (cp + 1, TSUBCWF);
		    if (fcc)
			fcc = add (", ", fcc);
		    fcc = add (cp, fcc);
                    free(dp);
		    continue;

		case FILESW: 
		    if (file)
			die("only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FILTSW:
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    filter = mh_xstrdup(etcpath(cp));
		    mime = false;
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			die("missing argument to %s", argp[-2]);
		    continue;

		case FRMTSW: 
		    filter = mh_xstrdup(etcpath(mhlreply));
		    mime = false;
		    continue;
		case NFRMTSW: 
		    filter = NULL;
		    continue;

		case INPLSW: 
		    inplace = true;
		    continue;
		case NINPLSW: 
		    inplace = false;
		    continue;

		case MIMESW:
		    mime = true;
		    filter = NULL;
		    continue;
		case NMIMESW:
		    mime = false;
		    continue;

		case QURYSW: 
		    querysw = true;
		    continue;
		case NQURYSW: 
		    querysw = false;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    if ((outputlinelen = atoi (cp)) < 10)
			die("impossible width %d", outputlinelen);
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			die("only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			die("missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW:
		    if (dmsg)
			die("only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			die("missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case ATFILESW:
		    atfile = true;
		    continue;
		case NOATFILESW:
		    atfile = false;
		    continue;

		case FMTPROCSW:
		    if (!(formatproc = *argp++) || *formatproc == '-')
			die("missing argument to %s", argp[-2]);
		    fmtproc = 1;
		    continue;
		case NFMTPROCSW:
		    fmtproc = 0;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		die("only one folder at a time!");
            folder = pluspath (cp);
	} else {
	    if (msg)
		die("only one message at a time!");
            msg = cp;
	}
    }

    if (ccto == -1) 
	ccto = groupreply;
    if (cccc == -1)
	cccc = groupreply;
    if (ccme == -1)
	ccme = groupreply;

    cwd = mh_xstrdup(pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msg || folder))
	die("can't mix files and folders/msgs");

try_it_again:

    strncpy (drft, buildsw ? m_maildir ("reply")
             : m_draft (dfolder, NULL, NOUSE, &isdf), sizeof(drft) - 1);

    /* Check if a draft exists */
    if (!buildsw && stat (drft, &st) != -1) {
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = read_switch_multiword ("\nDisposition? ",
						isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    inform("say what?");
		    break;
	    }
	}
    }

    if (file) {
	/*
	 * We are replying to a file.
	 */
	anot = false;	/* we don't want to annotate a file */
    } else {
	/*
	 * We are replying to a message.
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == -1)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder, 1)))
	    die("unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    die("no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (mp->numsel > 1)
	    die("only one message at a time!");

	context_replace (pfolder, folder);	/* update current folder   */
	seq_setcur (mp, mp->lowsel);	/* update current message  */
	seq_save (mp);			/* synchronize sequences   */
	context_save ();			/* save the context file   */
    }

    msg = file ? file : mh_xstrdup(m_name (mp->lowsel));

    if ((in = fopen (msg, "r")) == NULL)
	adios (msg, "unable to open");

    /* find form (components) file */
    if (!form) {
	if (groupreply)
	    form = etcpath (replgroupcomps);
	else
	    form = etcpath (replcomps);
    }

    replout (in, msg, drft, mp, outputlinelen, mime, form, filter,
	     fcc, fmtproc);
    fclose (in);

    {
        char *filename = file ? file : concat (mp->foldpath, "/", msg, NULL);

        for (n = 0; n < svector_size (convert_types); ++n) {
            add_convert_header (svector_at (convert_types, n),
                                svector_at (convert_args, n),
                                filename, drft);
        }
        if (! file) {
            free (filename);
        }
    }

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, msg, 0, mp, anot ? "Replied" : NULL,
	    inplace, cwd, atfile);

    svector_free (convert_args);
    svector_free (convert_types);

    done (1);
    return 1;
}

static void
docc (char *cp, int ccflag)
{
    switch (smatch (cp, ccswitches)) {
	case AMBIGSW: 
	    ambigsw (cp, ccswitches);
	    done (1);
	case UNKWNSW: 
	    die("-%scc %s unknown", ccflag ? "" : "no", cp);

	case CTOSW: 
	    ccto = ccflag;
	    break;

	case CCCSW: 
	    cccc = ccflag;
	    break;

	case CMESW: 
	    ccme = ccflag;
	    break;

	case CALSW: 
	    ccto = cccc = ccme = ccflag;
	    break;
    }
}

/*
 * Add pseudoheaders that will pass the convert arguments to
 * mhbuild.  They have the form:
 *   MHBUILD_FILE_PSEUDOHEADER-text/calendar: /home/user/Mail/inbox/7
 *   MHBUILD_ARGS_PSEUDOHEADER-text/calendar: reply -accept
 * The ARGS pseudoheader is optional, but we always add it when
 * -convertargs is used.
 */
static void
add_convert_header (const char *convert_type, char *convert_arg,
                    char *filename, char *drft)
{
    char *field_name;

    field_name = concat (MHBUILD_FILE_PSEUDOHEADER, convert_type, NULL);
    annotate (drft, field_name, filename, 1, 0, -2, 1);
    free (field_name);

    field_name = concat (MHBUILD_ARGS_PSEUDOHEADER, convert_type, NULL);
    annotate (drft, field_name, convert_arg, 1, 0, -2, 1);
    free (field_name);
}
